datafusion/optimizer/src/decorrelate_where_in.rs

```diff
@@ -121,80 +121,95 @@ impl OptimizerRule for DecorrelateWhereIn {
+    let subquery_input = projection.input.clone();
+    let subquery_expr = only_or_err(projection.expr.as_slice())
         .map_err(|e| context!("single expression projection required", e))?;
-    let subquery_col = proj
-        .try_into_col()
-        .map_err(|e| context!("single column projection required", e))?;
-    let outer_col = query_info
-        .where_in_expr
-        .try_into_col()
-        .map_err(|e| context!("column comparison required", e))?;
-
-    // If subquery is correlated, grab necessary information
-    let mut subqry_cols = vec![];
-    let mut outer_cols = vec![];
-    let mut join_filters = None;
-    let mut other_subqry_exprs = vec![];
-    if let LogicalPlan::Filter(subqry_filter) = (*subqry_input).clone() {
-        // split into filters
-        let subqry_filter_exprs = split_conjunction(&subqry_filter.predicate);
-        verify_not_disjunction(&subqry_filter_exprs)?;
-
-        // Grab column names to join on
-        let (col_exprs, other_exprs) =
-            find_join_exprs(subqry_filter_exprs, subqry_filter.input.schema())
-                .map_err(|e| context!("column correlation not found", e))?;
-        if !col_exprs.is_empty() {
-            // it's correlated
-            subqry_input = subqry_filter.input.clone();
-            (outer_cols, subqry_cols, join_filters) =
-                exprs_to_join_cols(&col_exprs, subqry_filter.input.schema(), false)
-                    .map_err(|e| context!("column correlation not found", e))?;
-            other_subqry_exprs = other_exprs;
-        }
-    }
 
-    let (subqry_cols, outer_cols) =
-        merge_cols((&[subquery_col], &subqry_cols), (&[outer_col], &outer_cols));
-
-    // build subquery side of join - the thing the subquery was querying
-    let subqry_alias = alias.next("__correlated_sq");
-    let mut subqry_plan = LogicalPlanBuilder::from((*subqry_input).clone());
-    if let Some(expr) = conjunction(other_subqry_exprs) {
-        // if the subquery had additional expressions, restore them
-        subqry_plan = subqry_plan.filter(expr)?
+    // extract join filters
+    let (join_filters, subquery_input) = extract_join_filters(subquery_input.as_ref())?;
+
+    // in_predicate may be also include in the join filters, remove it from the join filters.
+    let in_predicate = Expr::eq(query_info.where_in_expr.clone(), subquery_expr.clone());
+    let join_filters = remove_duplicated_filter(join_filters, in_predicate);
+
+    // replace qualified name with subquery alias.
+    let subquery_alias = alias.next("__correlated_sq");
+    let input_schema = subquery_input.schema();
+    let mut subquery_cols: BTreeSet<Column> =
+        join_filters
+            .iter()
+            .try_fold(BTreeSet::<Column>::new(), |mut cols, expr| {
+                let using_cols: Vec<Column> = expr
+                    .to_columns()?
+                    .into_iter()
+                    .filter(|col| input_schema.field_from_column(col).is_ok())
```

Comment:
doesn't this ignore columns that are not found in the subquery? I wonder if the error should be percolated up rather than ignored? Perhaps test coverage of a query like `select ... from t where x in (select non_existent_col from foo)` would be good?

Although perhaps if it was an error it signals something is wrong in the plan (as a column reference wouldn't be present in the input schema 🤔 )

If this shouldn't happen, maybe we can collect the errors and return them?

Reply:
> doesn't this ignore columns that are not found in the subquery

Yes, but it is used to collect columns which will be add to subquery projection.

Suppose there is a query:
`select * from t where x in (select c from foo where foo.a > t.a and foo.b > t.a)`

`foo.a > t.a and foo.b > t.a` will move to `join`, to make the join work, we need add `foo.a` and `foo.b` to the projection of subquery, the above code does the filter work.
 
 If the where clause reference unknown columns, we need do more work here, but I think it should be done by the binder.

Reply:
ok